optaplanner-core/src/main/java/org/optaplanner/core/impl/score/director/stream/DroolsConstraintStreamScoreDirectorFactory.java

```diff
@@ -180,36 +185,22 @@ public boolean isDroolsAlphaNetworkCompilationEnabled() {
         return droolsAlphaNetworkCompilationEnabled;
     }
 
-    private final class ConstraintDisablingAgendaFilter implements AgendaFilter {
+    private static final class ConstraintDisablingAgendaFilter implements AgendaFilter {
 
         private final Set<String> disabledConstraintIdSet;
+        private final Map<Rule, Boolean> ruleEnabledCache = new HashMap<>();
 
-        public ConstraintDisablingAgendaFilter(Map<DroolsConstraint<Solution_>, Score_> constraintToWeightMap) {
-            this.disabledConstraintIdSet = constraintToWeightMap.entrySet()
-                    .stream()
-                    .filter(entry -> Objects.equals(entry.getValue(), zeroScore))
-                    .map(Map.Entry::getKey)
-                    .map(Constraint::getConstraintId)
-                    .collect(toSet());
+        public ConstraintDisablingAgendaFilter(Set<String> disabledConstraintIdSet) {
+            this.disabledConstraintIdSet = Objects.requireNonNull(disabledConstraintIdSet);
         }
 
         @Override
         public boolean accept(Match match) {
-            if (disabledConstraintIdSet.isEmpty()) {
-                return true;
-            }
             Rule rule = match.getRule();
-            /*
-             * We identify the rule by its constraint ID, which we pre-calculated during rule creation.
-             * The alternative is to pay string concat penalty (packageName + name) to calculate the ID on every match.
-             * Since this code is on the hot path, this optimization was confirmed to bring considerable benefits.
-             */
-            String constraintId = (String) Objects.requireNonNull(
-                    rule.getMetaData().get(CONSTRAINT_ID_RULE_METADATA_KEY),
-                    () -> "Impossible state: Rule ("
-                            + ConstraintMatchTotal.composeConstraintId(rule.getPackageName(), rule.getName())
-                            + ") has no constraint ID.");
-            return !disabledConstraintIdSet.contains(constraintId);
+            return ruleEnabledCache.computeIfAbsent(rule, r -> {
+                String constraintId = ConstraintMatchTotal.composeConstraintId(r.getPackageName(), r.getName());
+                return !disabledConstraintIdSet.contains(constraintId);
+            });
```

Comment:
Reading 2 collections becomes reading just one collection, plus 1 write during cold start.
This should speed up this code considerably.